File: android/src/main/java/com/cloudwebrtc/webrtc/facedetection/FaceDetectionFrameProcessor.java
package com.cloudwebrtc.webrtc.facedetection;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ImageFormat;
import android.graphics.Matrix;
import android.graphics.Rect;
import android.graphics.YuvImage;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.Base64;
import android.util.Log;

import androidx.annotation.NonNull;

import com.cloudwebrtc.webrtc.video.LocalVideoTrack;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.face.Face;
import com.google.mlkit.vision.face.FaceDetection;
import com.google.mlkit.vision.face.FaceDetector;
import com.google.mlkit.vision.face.FaceDetectorOptions;
import com.google.mlkit.vision.face.FaceLandmark;

import org.webrtc.VideoFrame;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.flutter.plugin.common.EventChannel;

/**
 * Processes video frames for face detection using ML Kit.
 * Implements LocalVideoTrack.ExternalVideoFrameProcessing to receive frames from WebRTC.
 */
public class FaceDetectionFrameProcessor implements LocalVideoTrack.ExternalVideoFrameProcessing {
    private static final String TAG = "FaceDetection";

    private final FaceDetector faceDetector;
    private final EyeStateTracker eyeStateTracker;
    private final Handler processingHandler;
    private final HandlerThread processingThread;

    private EventChannel.EventSink faceEventSink;
    private EventChannel.EventSink blinkEventSink;

    private FaceDetectionConfig config;
    private int frameCount = 0;
    private volatile boolean isProcessing = false;
    private volatile boolean isDisposed = false;

    public FaceDetectionFrameProcessor() {
        // Configure ML Kit face detector
        FaceDetectorOptions options = new FaceDetectorOptions.Builder()
                .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
                .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
                .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
                .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
                .setMinFaceSize(0.15f)
                .enableTracking()
                .build();

        faceDetector = FaceDetection.getClient(options);
        eyeStateTracker = new EyeStateTracker();
        config = new FaceDetectionConfig();

        // Create dedicated processing thread
        processingThread = new HandlerThread("FaceDetectionThread");
        processingThread.start();
        processingHandler = new Handler(processingThread.getLooper());

        Log.d(TAG, "FaceDetectionFrameProcessor initialized");
    }

    public void setConfig(FaceDetectionConfig config) {
        this.config = config;
        eyeStateTracker.setBlinkThreshold(config.blinkThreshold);
    }

    public void setFaceEventSink(EventChannel.EventSink sink) {
        this.faceEventSink = sink;
    }

    public void setBlinkEventSink(EventChannel.EventSink sink) {
        this.blinkEventSink = sink;
    }

    @Override
    public VideoFrame onFrame(VideoFrame frame) {
        if (isDisposed) {
            return frame;
        }

        // Frame skip check
        frameCount++;
        if (frameCount % config.frameSkipCount != 0) {
            return frame;
        }

        // Non-blocking check - skip if still processing previous frame
        if (isProcessing) {
            return frame;
        }
        isProcessing = true;

        // Capture frame data before async processing
        final int width = frame.getBuffer().getWidth();
        final int height = frame.getBuffer().getHeight();
        final int rotation = frame.getRotation();
        final long timestampNs = frame.getTimestampNs();

        // Convert to I420 and copy data
        VideoFrame.I420Buffer i420Buffer = frame.getBuffer().toI420();
        final byte[] nv21Data = convertI420ToNV21(i420Buffer, width, height);
        i420Buffer.release();

        // Process asynchronously
        processingHandler.post(() -> {
            if (isDisposed) {
                isProcessing = false;
                return;
            }
            processFrameAsync(nv21Data, width, height, rotation, timestampNs);
        });

        return frame; // Return original frame for rendering
    }

    private void processFrameAsync(byte[] nv21Data, int width, int height, int rotation, long timestampNs) {
        try {
            // Create InputImage from NV21 data
            InputImage inputImage = InputImage.fromByteArray(
                    nv21Data,
                    width,
                    height,
                    rotation,
                    InputImage.IMAGE_FORMAT_NV21
            );

            // Run face detection
            faceDetector.process(inputImage)
                    .addOnSuccessListener(faces -> {
                        if (isDisposed) {
                            isProcessing = false;
                            return;
                        }
                        processFaceResults(faces, nv21Data, width, height, rotation, timestampNs);
                        isProcessing = false;
                    })
                    .addOnFailureListener(e -> {
                        Log.e(TAG, "Face detection failed", e);
                        isProcessing = false;
                    });

        } catch (Exception e) {
            Log.e(TAG, "Error processing frame", e);
            isProcessing = false;
        }
    }

    private void processFaceResults(List<Face> faces, byte[] nv21Data, int width, int height, int rotation, long timestampNs) {
        if (faceEventSink == null && blinkEventSink == null) {
            return;
        }

        Set<Integer> activeTrackingIds = new HashSet<>();
        List<Map<String, Object>> faceDataList = new ArrayList<>();

        for (Face face : faces) {
            Integer trackingId = face.getTrackingId();
            if (trackingId != null) {
                activeTrackingIds.add(trackingId);
            }

            Map<String, Object> faceData = new HashMap<>();

            // Bounds
            Rect bounds = face.getBoundingBox();
            Map<String, Object> boundsMap = new HashMap<>();
            boundsMap.put("left", bounds.left);
            boundsMap.put("top", bounds.top);
            boundsMap.put("width", bounds.width());
            boundsMap.put("height", bounds.height());
            faceData.put("bounds", boundsMap);

            // Tracking ID
            if (trackingId != null) {
                faceData.put("trackingId", trackingId);
            }

            // Head pose
            Map<String, Object> headPose = new HashMap<>();
            headPose.put("yaw", face.getHeadEulerAngleY());
            headPose.put("pitch", face.getHeadEulerAngleX());
            headPose.put("roll", face.getHeadEulerAngleZ());
            faceData.put("headPose", headPose);

            // Landmarks
            Map<String, Object> landmarks = new HashMap<>();

            // Left eye
            FaceLandmark leftEye = face.getLandmark(FaceLandmark.LEFT_EYE);
            if (leftEye != null) {
                Map<String, Object> leftEyeData = new HashMap<>();
                leftEyeData.put("x", leftEye.getPosition().x);
                leftEyeData.put("y", leftEye.getPosition().y);
                Float leftEyeOpenProb = face.getLeftEyeOpenProbability();
                if (leftEyeOpenProb != null) {
                    leftEyeData.put("openProbability", leftEyeOpenProb);
                    leftEyeData.put("isOpen", leftEyeOpenProb > config.blinkThreshold);
                }
                landmarks.put("leftEye", leftEyeData);
            }

            // Right eye
            FaceLandmark rightEye = face.getLandmark(FaceLandmark.RIGHT_EYE);
            if (rightEye != null) {
                Map<String, Object> rightEyeData = new HashMap<>();
                rightEyeData.put("x", rightEye.getPosition().x);
                rightEyeData.put("y", rightEye.getPosition().y);
                Float rightEyeOpenProb = face.getRightEyeOpenProbability();
                if (rightEyeOpenProb != null) {
                    rightEyeData.put("openProbability", rightEyeOpenProb);
                    rightEyeData.put("isOpen", rightEyeOpenProb > config.blinkThreshold);
                }
                landmarks.put("rightEye", rightEyeData);
            }

            // Nose
            FaceLandmark nose = face.getLandmark(FaceLandmark.NOSE_BASE);
            if (nose != null) {
                Map<String, Object> noseData = new HashMap<>();
                noseData.put("x", nose.getPosition().x);
                noseData.put("y", nose.getPosition().y);
                landmarks.put("nose", noseData);
            }

            // Mouth
            FaceLandmark mouthLeft = face.getLandmark(FaceLandmark.MOUTH_LEFT);
            FaceLandmark mouthRight = face.getLandmark(FaceLandmark.MOUTH_RIGHT);
            FaceLandmark mouthBottom = face.getLandmark(FaceLandmark.MOUTH_BOTTOM);
            if (mouthLeft != null && mouthRight != null) {
                Map<String, Object> mouthData = new HashMap<>();
                mouthData.put("leftX", mouthLeft.getPosition().x);
                mouthData.put("leftY", mouthLeft.getPosition().y);
                mouthData.put("rightX", mouthRight.getPosition().x);
                mouthData.put("rightY", mouthRight.getPosition().y);
                if (mouthBottom != null) {
                    mouthData.put("bottomX", mouthBottom.getPosition().x);
                    mouthData.put("bottomY", mouthBottom.getPosition().y);
                }
                Float smilingProb = face.getSmilingProbability();
                if (smilingProb != null) {
                    mouthData.put("smilingProbability", smilingProb);
                }
                landmarks.put("mouth", mouthData);
            }

            faceData.put("landmarks", landmarks);

            // Smiling probability
            Float smilingProb = face.getSmilingProbability();
            if (smilingProb != null) {
                faceData.put("smilingProbability", smilingProb);
            }

            faceDataList.add(faceData);

            // Blink detection
            if (trackingId != null && blinkEventSink != null) {
                Float leftEyeOpenProb = face.getLeftEyeOpenProbability();
                Float rightEyeOpenProb = face.getRightEyeOpenProbability();

                if (leftEyeOpenProb != null && rightEyeOpenProb != null) {
                    // Capture frame if configured and eye is closing
                    String capturedFrame = null;
                    if (config.captureOnBlink) {
                        boolean leftClosing = leftEyeOpenProb <= config.blinkThreshold;
                        boolean rightClosing = rightEyeOpenProb <= config.blinkThreshold;
                        EyeStateTracker.FaceEyeState faceState = eyeStateTracker.getFaceState(trackingId);
                        boolean wasLeftOpen = faceState == null || faceState.leftEye.isOpen;
                        boolean wasRightOpen = faceState == null || faceState.rightEye.isOpen;

                        if ((leftClosing && wasLeftOpen) || (rightClosing && wasRightOpen)) {
                            capturedFrame = captureFrameAsBase64(nv21Data, width, height, bounds);
                        }
                    }

                    EyeStateTracker.BlinkResult blinkResult = eyeStateTracker.updateEyeState(
                            trackingId,
                            leftEyeOpenProb,
                            rightEyeOpenProb,
                            capturedFrame
                    );

                    if (blinkResult != null) {
                        Map<String, Object> blinkEvent = blinkResult.toMap();
                        blinkEvent.put("trackingId", trackingId);
                        blinkEvent.put("timestamp", timestampNs);
                        emitBlinkEvent(blinkEvent);
                    }
                }
            }
        }

        // Cleanup stale face states
        eyeStateTracker.cleanupStaleStates(activeTrackingIds);

        // Emit face detection results
        if (faceEventSink != null) {
            Map<String, Object> result = new HashMap<>();
            result.put("faces", faceDataList);
            result.put("timestamp", timestampNs);
            result.put("frameWidth", width);
            result.put("frameHeight", height);
            emitFaceEvent(result);
        }
    }

    private void emitFaceEvent(Map<String, Object> event) {
        if (faceEventSink != null) {
            try {
                // Must emit on main thread
                new Handler(android.os.Looper.getMainLooper()).post(() -> {
                    if (faceEventSink != null && !isDisposed) {
                        faceEventSink.success(event);
                    }
                });
            } catch (Exception e) {
                Log.e(TAG, "Error emitting face event", e);
            }
        }
    }

    private void emitBlinkEvent(Map<String, Object> event) {
        if (blinkEventSink != null) {
            try {
                // Must emit on main thread
                new Handler(android.os.Looper.getMainLooper()).post(() -> {
                    if (blinkEventSink != null && !isDisposed) {
                        blinkEventSink.success(event);
                    }
                });
            } catch (Exception e) {
                Log.e(TAG, "Error emitting blink event", e);
            }
        }
    }

    private String captureFrameAsBase64(byte[] nv21Data, int width, int height, Rect faceBounds) {
        try {
            // Convert NV21 to JPEG
            YuvImage yuvImage = new YuvImage(nv21Data, ImageFormat.NV21, width, height, null);
            ByteArrayOutputStream jpegStream = new ByteArrayOutputStream();

            Rect cropRect;
            if (config.cropToFace && faceBounds != null) {
                // Expand face bounds slightly for context
                int padding = (int) (Math.min(faceBounds.width(), faceBounds.height()) * 0.2);
                cropRect = new Rect(
                        Math.max(0, faceBounds.left - padding),
                        Math.max(0, faceBounds.top - padding),
                        Math.min(width, faceBounds.right + padding),
                        Math.min(height, faceBounds.bottom + padding)
                );
            } else {
                cropRect = new Rect(0, 0, width, height);
            }

            int quality = (int) (config.imageQuality * 100);
            yuvImage.compressToJpeg(cropRect, quality, jpegStream);

            byte[] jpegBytes = jpegStream.toByteArray();

            // Resize if necessary
            if (config.maxImageWidth > 0 && cropRect.width() > config.maxImageWidth) {
                Bitmap bitmap = BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.length);
                float scale = (float) config.maxImageWidth / bitmap.getWidth();
                int newHeight = (int) (bitmap.getHeight() * scale);

                Bitmap resized = Bitmap.createScaledBitmap(bitmap, config.maxImageWidth, newHeight, true);
                bitmap.recycle();

                ByteArrayOutputStream resizedStream = new ByteArrayOutputStream();
                resized.compress(Bitmap.CompressFormat.JPEG, quality, resizedStream);
                resized.recycle();

                jpegBytes = resizedStream.toByteArray();
            }

            return Base64.encodeToString(jpegBytes, Base64.NO_WRAP);
        } catch (Exception e) {
            Log.e(TAG, "Error capturing frame", e);
            return null;
        }
    }

    /**
     * Convert I420 buffer to NV21 byte array.
     */
    private byte[] convertI420ToNV21(VideoFrame.I420Buffer i420Buffer, int width, int height) {
        int chromaWidth = (width + 1) / 2;
        int chromaHeight = (height + 1) / 2;

        int ySize = width * height;
        int uvSize = chromaWidth * chromaHeight * 2;
        byte[] nv21 = new byte[ySize + uvSize];

        // Copy Y plane
        ByteBuffer yBuffer = i420Buffer.getDataY();
        int yStride = i420Buffer.getStrideY();
        for (int row = 0; row < height; row++) {
            yBuffer.position(row * yStride);
            yBuffer.get(nv21, row * width, width);
        }

        // Interleave U and V planes to VU (NV21 format)
        ByteBuffer uBuffer = i420Buffer.getDataU();
        ByteBuffer vBuffer = i420Buffer.getDataV();
        int uStride = i420Buffer.getStrideU();
        int vStride = i420Buffer.getStrideV();

        int uvOffset = ySize;
        for (int row = 0; row < chromaHeight; row++) {
            for (int col = 0; col < chromaWidth; col++) {
                int uIndex = row * uStride + col;
                int vIndex = row * vStride + col;
                nv21[uvOffset++] = vBuffer.get(vIndex); // V first in NV21
                nv21[uvOffset++] = uBuffer.get(uIndex); // then U
            }
        }

        return nv21;
    }

    public void dispose() {
        isDisposed = true;
        isProcessing = false;

        processingHandler.post(() -> {
            faceDetector.close();
            eyeStateTracker.reset();
        });

        processingThread.quitSafely();

        faceEventSink = null;
        blinkEventSink = null;

        Log.d(TAG, "FaceDetectionFrameProcessor disposed");
    }
}

File: common/darwin/Classes/FaceDetectionFrameProcessor.m
#import "FaceDetectionFrameProcessor.h"
#import <CoreVideo/CoreVideo.h>

@import MLKitFaceDetection;
@import MLKitVision;

#pragma mark - FaceDetectionConfig

@implementation FaceDetectionConfig

- (instancetype)init {
    self = [super init];
    if (self) {
        _frameSkipCount = 3;
        _blinkThreshold = 0.3;
        _captureOnBlink = NO;
        _cropToFace = YES;
        _imageQuality = 0.7;
        _maxImageWidth = 480;
    }
    return self;
}

+ (instancetype)configFromDictionary:(NSDictionary*)dict {
    FaceDetectionConfig* config = [[FaceDetectionConfig alloc] init];
    if (dict == nil) {
        return config;
    }

    if (dict[@"frameSkipCount"]) {
        config.frameSkipCount = [dict[@"frameSkipCount"] integerValue];
    }
    if (dict[@"blinkThreshold"]) {
        config.blinkThreshold = [dict[@"blinkThreshold"] doubleValue];
    }
    if (dict[@"captureOnBlink"]) {
        config.captureOnBlink = [dict[@"captureOnBlink"] boolValue];
    }
    if (dict[@"cropToFace"]) {
        config.cropToFace = [dict[@"cropToFace"] boolValue];
    }
    if (dict[@"imageQuality"]) {
        config.imageQuality = [dict[@"imageQuality"] doubleValue];
    }
    if (dict[@"maxImageWidth"]) {
        config.maxImageWidth = [dict[@"maxImageWidth"] integerValue];
    }

    return config;
}

@end

#pragma mark - EyeState

@implementation EyeState

- (instancetype)init {
    self = [super init];
    if (self) {
        _wasOpen = YES;
        _isOpen = YES;
        _blinkCount = 0;
        _pendingCapturedFrame = nil;
    }
    return self;
}

@end

#pragma mark - FaceEyeState

@implementation FaceEyeState

- (instancetype)init {
    self = [super init];
    if (self) {
        _leftEye = [[EyeState alloc] init];
        _rightEye = [[EyeState alloc] init];
    }
    return self;
}

@end

#pragma mark - EyeStateTracker

@interface EyeStateTracker ()
@property(nonatomic, strong) NSMutableDictionary<NSNumber*, FaceEyeState*>* faceStates;
@property(nonatomic, assign) double blinkThreshold;
@end

@implementation EyeStateTracker

- (instancetype)init {
    self = [super init];
    if (self) {
        _faceStates = [NSMutableDictionary new];
        _blinkThreshold = 0.3;
    }
    return self;
}

- (void)setBlinkThreshold:(double)threshold {
    _blinkThreshold = threshold;
}

- (NSDictionary*)updateEyeStateForTrackingId:(NSInteger)trackingId
                             leftEyeOpenProb:(float)leftProb
                            rightEyeOpenProb:(float)rightProb
                               capturedFrame:(NSString*)capturedFrame {
    NSNumber* key = @(trackingId);
    FaceEyeState* faceState = _faceStates[key];
    if (faceState == nil) {
        faceState = [[FaceEyeState alloc] init];
        _faceStates[key] = faceState;
    }

    BOOL leftCurrentlyOpen = leftProb > _blinkThreshold;
    BOOL rightCurrentlyOpen = rightProb > _blinkThreshold;

    BOOL leftBlinked = NO;
    BOOL rightBlinked = NO;
    NSString* blinkCapturedFrame = nil;

    // Check left eye
    if (!faceState.leftEye.isOpen && leftCurrentlyOpen) {
        leftBlinked = YES;
        faceState.leftEye.blinkCount++;
        blinkCapturedFrame = faceState.leftEye.pendingCapturedFrame;
        faceState.leftEye.pendingCapturedFrame = nil;
    } else if (faceState.leftEye.isOpen && !leftCurrentlyOpen) {
        faceState.leftEye.pendingCapturedFrame = capturedFrame;
    }

    // Check right eye
    if (!faceState.rightEye.isOpen && rightCurrentlyOpen) {
        rightBlinked = YES;
        faceState.rightEye.blinkCount++;
        if (blinkCapturedFrame == nil) {
            blinkCapturedFrame = faceState.rightEye.pendingCapturedFrame;
        }
        faceState.rightEye.pendingCapturedFrame = nil;
    } else if (faceState.rightEye.isOpen && !rightCurrentlyOpen) {
        faceState.rightEye.pendingCapturedFrame = capturedFrame;
    }

    // Update state
    faceState.leftEye.wasOpen = faceState.leftEye.isOpen;
    faceState.leftEye.isOpen = leftCurrentlyOpen;
    faceState.rightEye.wasOpen = faceState.rightEye.isOpen;
    faceState.rightEye.isOpen = rightCurrentlyOpen;

    if (leftBlinked || rightBlinked) {
        NSString* eye;
        if (leftBlinked && rightBlinked) {
            eye = @"both";
        } else if (leftBlinked) {
            eye = @"left";
        } else {
            eye = @"right";
        }

        NSMutableDictionary* result = [NSMutableDictionary new];
        result[@"eye"] = eye;
        result[@"leftBlinkCount"] = @(faceState.leftEye.blinkCount);
        result[@"rightBlinkCount"] = @(faceState.rightEye.blinkCount);
        if (blinkCapturedFrame) {
            result[@"capturedFrame"] = blinkCapturedFrame;
        }
        return result;
    }

    return nil;
}

- (FaceEyeState*)getFaceState:(NSInteger)trackingId {
    return _faceStates[@(trackingId)];
}

- (void)cleanupStaleStates:(NSSet<NSNumber*>*)activeTrackingIds {
    NSMutableArray* keysToRemove = [NSMutableArray new];
    for (NSNumber* key in _faceStates) {
        if (![activeTrackingIds containsObject:key]) {
            [keysToRemove addObject:key];
        }
    }
    [_faceStates removeObjectsForKeys:keysToRemove];
}

- (void)reset {
    [_faceStates removeAllObjects];
}

@end

#pragma mark - FaceDetectionFrameProcessor

@interface FaceDetectionFrameProcessor ()
@property(nonatomic, strong) MLKFaceDetector* faceDetector;
@property(nonatomic, strong) EyeStateTracker* eyeStateTracker;
@property(nonatomic, strong) dispatch_queue_t processingQueue;
@property(nonatomic, assign) NSInteger frameCount;
@property(nonatomic, assign) BOOL isProcessing;
@property(nonatomic, assign) BOOL isDisposed;
@end

@implementation FaceDetectionFrameProcessor

- (instancetype)init {
    self = [super init];
    if (self) {
        // Configure ML Kit face detector
        MLKFaceDetectorOptions* options = [[MLKFaceDetectorOptions alloc] init];
        options.performanceMode = MLKFaceDetectorPerformanceModeFast;
        options.landmarkMode = MLKFaceDetectorLandmarkModeAll;
        options.classificationMode = MLKFaceDetectorClassificationModeAll;
        options.contourMode = MLKFaceDetectorContourModeNone;
        options.minFaceSize = 0.15;
        options.trackingEnabled = YES;

        _faceDetector = [MLKFaceDetector faceDetectorWithOptions:options];
        _eyeStateTracker = [[EyeStateTracker alloc] init];
        _config = [[FaceDetectionConfig alloc] init];
        _processingQueue = dispatch_queue_create("com.cloudwebrtc.facedetection", DISPATCH_QUEUE_SERIAL);
        _frameCount = 0;
        _isProcessing = NO;
        _isDisposed = NO;

        NSLog(@"FaceDetectionFrameProcessor initialized");
    }
    return self;
}

- (void)setConfig:(FaceDetectionConfig*)config {
    _config = config;
    [_eyeStateTracker setBlinkThreshold:config.blinkThreshold];
}

- (RTCVideoFrame*)onFrame:(RTCVideoFrame*)frame {
    if (_isDisposed) {
        return frame;
    }

    // Frame skip check
    _frameCount++;
    if (_frameCount % _config.frameSkipCount != 0) {
        return frame;
    }

    // Non-blocking check
    if (_isProcessing) {
        return frame;
    }
    _isProcessing = YES;

    // Retain frame data for async processing
    int64_t timestampNs = frame.timeStampNs;
    int width = frame.width;
    int height = frame.height;
    int rotation = frame.rotation;

    // Get the pixel buffer
    id<RTCVideoFrameBuffer> buffer = frame.buffer;
    CVPixelBufferRef pixelBuffer = NULL;

    if ([buffer isKindOfClass:[RTCCVPixelBuffer class]]) {
        RTCCVPixelBuffer* cvPixelBuffer = (RTCCVPixelBuffer*)buffer;
        pixelBuffer = cvPixelBuffer.pixelBuffer;
        CVPixelBufferRetain(pixelBuffer);
    }

    if (pixelBuffer == NULL) {
        _isProcessing = NO;
        return frame;
    }

    dispatch_async(_processingQueue, ^{
        if (self.isDisposed) {
            CVPixelBufferRelease(pixelBuffer);
            self.isProcessing = NO;
            return;
        }

        [self processPixelBuffer:pixelBuffer
                           width:width
                          height:height
                        rotation:rotation
                     timestampNs:timestampNs];

        CVPixelBufferRelease(pixelBuffer);
    });

    return frame;
}

- (void)processPixelBuffer:(CVPixelBufferRef)pixelBuffer
                     width:(int)width
                    height:(int)height
                  rotation:(int)rotation
               timestampNs:(int64_t)timestampNs {

    UIImageOrientation orientation;
    switch (rotation) {
        case 90:
            orientation = UIImageOrientationRight;
            break;
        case 180:
            orientation = UIImageOrientationDown;
            break;
        case 270:
            orientation = UIImageOrientationLeft;
            break;
        default:
            orientation = UIImageOrientationUp;
            break;
    }

    MLKVisionImage* visionImage = [[MLKVisionImage alloc] initWithBuffer:pixelBuffer];
    visionImage.orientation = orientation;

    [_faceDetector processImage:visionImage
                     completion:^(NSArray<MLKFace*>* _Nullable faces, NSError* _Nullable error) {
        if (self.isDisposed) {
            self.isProcessing = NO;
            return;
        }

        if (error) {
            NSLog(@"Face detection failed: %@", error.localizedDescription);
            self.isProcessing = NO;
            return;
        }

        [self processFaceResults:faces
                     pixelBuffer:pixelBuffer
                           width:width
                          height:height
                        rotation:rotation
                     timestampNs:timestampNs];

        self.isProcessing = NO;
    }];
}

- (void)processFaceResults:(NSArray<MLKFace*>*)faces
               pixelBuffer:(CVPixelBufferRef)pixelBuffer
                     width:(int)width
                    height:(int)height
                  rotation:(int)rotation
               timestampNs:(int64_t)timestampNs {

    if (_faceEventSink == nil && _blinkEventSink == nil) {
        return;
    }

    NSMutableSet<NSNumber*>* activeTrackingIds = [NSMutableSet new];
    NSMutableArray* faceDataList = [NSMutableArray new];

    for (MLKFace* face in faces) {
        NSInteger trackingId = face.trackingID;
        [activeTrackingIds addObject:@(trackingId)];

        NSMutableDictionary* faceData = [NSMutableDictionary new];

        // Bounds
        CGRect bounds = face.frame;
        faceData[@"bounds"] = @{
            @"left": @(bounds.origin.x),
            @"top": @(bounds.origin.y),
            @"width": @(bounds.size.width),
            @"height": @(bounds.size.height)
        };

        // Tracking ID
        if (face.hasTrackingID) {
            faceData[@"trackingId"] = @(trackingId);
        }

        // Head pose
        faceData[@"headPose"] = @{
            @"yaw": @(face.headEulerAngleY),
            @"pitch": @(face.headEulerAngleX),
            @"roll": @(face.headEulerAngleZ)
        };

        // Landmarks
        NSMutableDictionary* landmarks = [NSMutableDictionary new];

        // Left eye
        MLKFaceLandmark* leftEye = [face landmarkOfType:MLKFaceLandmarkTypeLeftEye];
        if (leftEye) {
            NSMutableDictionary* leftEyeData = [NSMutableDictionary new];
            leftEyeData[@"x"] = @(leftEye.position.x);
            leftEyeData[@"y"] = @(leftEye.position.y);
            if (face.hasLeftEyeOpenProbability) {
                leftEyeData[@"openProbability"] = @(face.leftEyeOpenProbability);
                leftEyeData[@"isOpen"] = @(face.leftEyeOpenProbability > _config.blinkThreshold);
            }
            landmarks[@"leftEye"] = leftEyeData;
        }

        // Right eye
        MLKFaceLandmark* rightEye = [face landmarkOfType:MLKFaceLandmarkTypeRightEye];
        if (rightEye) {
            NSMutableDictionary* rightEyeData = [NSMutableDictionary new];
            rightEyeData[@"x"] = @(rightEye.position.x);
            rightEyeData[@"y"] = @(rightEye.position.y);
            if (face.hasRightEyeOpenProbability) {
                rightEyeData[@"openProbability"] = @(face.rightEyeOpenProbability);
                rightEyeData[@"isOpen"] = @(face.rightEyeOpenProbability > _config.blinkThreshold);
            }
            landmarks[@"rightEye"] = rightEyeData;
        }

        // Nose
        MLKFaceLandmark* nose = [face landmarkOfType:MLKFaceLandmarkTypeNoseBase];
        if (nose) {
            landmarks[@"nose"] = @{
                @"x": @(nose.position.x),
                @"y": @(nose.position.y)
            };
        }

        // Mouth
        MLKFaceLandmark* mouthLeft = [face landmarkOfType:MLKFaceLandmarkTypeMouthLeft];
        MLKFaceLandmark* mouthRight = [face landmarkOfType:MLKFaceLandmarkTypeMouthRight];
        MLKFaceLandmark* mouthBottom = [face landmarkOfType:MLKFaceLandmarkTypeMouthBottom];
        if (mouthLeft && mouthRight) {
            NSMutableDictionary* mouthData = [NSMutableDictionary new];
            mouthData[@"leftX"] = @(mouthLeft.position.x);
            mouthData[@"leftY"] = @(mouthLeft.position.y);
            mouthData[@"rightX"] = @(mouthRight.position.x);
            mouthData[@"rightY"] = @(mouthRight.position.y);
            if (mouthBottom) {
                mouthData[@"bottomX"] = @(mouthBottom.position.x);
                mouthData[@"bottomY"] = @(mouthBottom.position.y);
            }
            if (face.hasSmilingProbability) {
                mouthData[@"smilingProbability"] = @(face.smilingProbability);
            }
            landmarks[@"mouth"] = mouthData;
        }

        faceData[@"landmarks"] = landmarks;

        // Smiling probability
        if (face.hasSmilingProbability) {
            faceData[@"smilingProbability"] = @(face.smilingProbability);
        }

        [faceDataList addObject:faceData];

        // Blink detection
        if (face.hasTrackingID && _blinkEventSink != nil) {
            if (face.hasLeftEyeOpenProbability && face.hasRightEyeOpenProbability) {
                NSString* capturedFrame = nil;

                if (_config.captureOnBlink) {
                    BOOL leftClosing = face.leftEyeOpenProbability <= _config.blinkThreshold;
                    BOOL rightClosing = face.rightEyeOpenProbability <= _config.blinkThreshold;
                    FaceEyeState* faceState = [_eyeStateTracker getFaceState:trackingId];
                    BOOL wasLeftOpen = faceState == nil || faceState.leftEye.isOpen;
                    BOOL wasRightOpen = faceState == nil || faceState.rightEye.isOpen;

                    if ((leftClosing && wasLeftOpen) || (rightClosing && wasRightOpen)) {
                        capturedFrame = [self captureFrameAsBase64:pixelBuffer
                                                            width:width
                                                           height:height
                                                       faceBounds:bounds];
                    }
                }

                NSDictionary* blinkResult = [_eyeStateTracker updateEyeStateForTrackingId:trackingId
                                                                          leftEyeOpenProb:face.leftEyeOpenProbability
                                                                         rightEyeOpenProb:face.rightEyeOpenProbability
                                                                            capturedFrame:capturedFrame];

                if (blinkResult) {
                    NSMutableDictionary* blinkEvent = [blinkResult mutableCopy];
                    blinkEvent[@"trackingId"] = @(trackingId);
                    blinkEvent[@"timestamp"] = @(timestampNs);
                    [self emitBlinkEvent:blinkEvent];
                }
            }
        }
    }

    // Cleanup stale face states
    [_eyeStateTracker cleanupStaleStates:activeTrackingIds];

    // Emit face detection results
    if (_faceEventSink != nil) {
        NSDictionary* result = @{
            @"faces": faceDataList,
            @"timestamp": @(timestampNs),
            @"frameWidth": @(width),
            @"frameHeight": @(height)
        };
        [self emitFaceEvent:result];
    }
}

- (void)emitFaceEvent:(NSDictionary*)event {
    if (_faceEventSink == nil || _isDisposed) {
        return;
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.faceEventSink != nil && !self.isDisposed) {
            self.faceEventSink(event);
        }
    });
}

- (void)emitBlinkEvent:(NSDictionary*)event {
    if (_blinkEventSink == nil || _isDisposed) {
        return;
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.blinkEventSink != nil && !self.isDisposed) {
            self.blinkEventSink(event);
        }
    });
}

- (NSString*)captureFrameAsBase64:(CVPixelBufferRef)pixelBuffer
                            width:(int)width
                           height:(int)height
                       faceBounds:(CGRect)faceBounds {
    @try {
        CIImage* ciImage = [CIImage imageWithCVPixelBuffer:pixelBuffer];

        CGRect cropRect;
        if (_config.cropToFace && !CGRectIsEmpty(faceBounds)) {
            CGFloat padding = MIN(faceBounds.size.width, faceBounds.size.height) * 0.2;
            cropRect = CGRectMake(
                MAX(0, faceBounds.origin.x - padding),
                MAX(0, faceBounds.origin.y - padding),
                MIN(width - faceBounds.origin.x + padding, faceBounds.size.width + padding * 2),
                MIN(height - faceBounds.origin.y + padding, faceBounds.size.height + padding * 2)
            );
        } else {
            cropRect = CGRectMake(0, 0, width, height);
        }

        ciImage = [ciImage imageByCroppingToRect:cropRect];

        CIContext* context = [CIContext contextWithOptions:nil];
        CGImageRef cgImage = [context createCGImage:ciImage fromRect:ciImage.extent];

        if (cgImage == NULL) {
            return nil;
        }

        UIImage* image = [UIImage imageWithCGImage:cgImage];
        CGImageRelease(cgImage);

        // Resize if necessary
        if (_config.maxImageWidth > 0 && image.size.width > _config.maxImageWidth) {
            CGFloat scale = _config.maxImageWidth / image.size.width;
            CGSize newSize = CGSizeMake(_config.maxImageWidth, image.size.height * scale);

            UIGraphicsBeginImageContextWithOptions(newSize, NO, 1.0);
            [image drawInRect:CGRectMake(0, 0, newSize.width, newSize.height)];
            image = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();
        }

        NSData* jpegData = UIImageJPEGRepresentation(image, _config.imageQuality);
        return [jpegData base64EncodedStringWithOptions:0];

    } @catch (NSException* exception) {
        NSLog(@"Error capturing frame: %@", exception.reason);
        return nil;
    }
}

- (void)dispose {
    _isDisposed = YES;
    _isProcessing = NO;
    [_eyeStateTracker reset];
    _faceEventSink = nil;
    _blinkEventSink = nil;

    NSLog(@"FaceDetectionFrameProcessor disposed");
}

@end
